_posts/2021/2021-10-04-use-sidekiq-directly-not-through-active-job.md

```diff
@@ -0,0 +1,84 @@
+description: "Active Job is great, but you might be missing out."
+layout: article
+category: ruby
+image:
+  base: '2021/use-sidekiq-directly-not-through-active-job'
+  alt: "Silhouetted karate kick on the beach"
+  credit: "Jason Briscoe"
+  source: "https://unsplash.com/photos/HN_4K2diUWs"
+
+---
+
+If you build a web application you should minimize the amount of time spent responding to every user; a fast website means a happy user. One way of doing that is to run long-running, parallelisable, or potentially slow, work asynchronously outside of the immediate web request. This might be sending emails, scheduled cleanups, long running calculations, or anything that uses an external API.
+
+[Active Job](https://guides.rubyonrails.org/active_job_basics.html) is the recommended way of enqueuing and running background work in Rails. It provides a wrapper around a bunch of [different adapter implementations](https://api.rubyonrails.org/classes/ActiveJob/QueueAdapters.html) of the concept. Each implementation utilises a different underlying technology and has its own pros and cons. Most of the implementations existed before [Active Job was released](https://guides.rubyonrails.org/4_2_release_notes.html#active-job) in Rails 4.2.
+
+One of the most popular, battle-tested, and performant asynchronous frameworks is [Sidekiq](https://sidekiq.org), which is multi-threaded and utilises Redis for its queuing storage. It’s an open source library with two paid tiers for additional high-end features.
+
+
+## Instead of…
+
+…defining your jobs using Active Job:
+
+```ruby
+class DoThingsInBackgroundJob < ApplicationJob
+  queue_as :default
+
+  def peform(an_active_record_object)
+    an_active_record_object.do_things
+  end
+end
+```
+
+## Use…
+
+…Sidekiq directly:
+
+```ruby
+class DoThingsInBackgroundJob
+  include Sidekiq::Worker
+  Sidekiq_options queue: "default"
+
+  def peform(id)
+    an_active_record_object = ActiveRecordObject.find_by(id: id)
+    an_active_record_object.do_things
+  end
+end
+```
+
+
+## Why?
+
+You’re missing out on a number of advantages that you gain when you use Sidekiq directly.
+
+If you're doing _a lot_ of work in the background, enqueuing many small fast-running jobs—as is generally suggested—there is a [2—20x performance improvement](https://github.com/mperham/sidekiq/wiki/Active-Job#performance) if you use Sidekiq directly. The level of improvement will be dependant on your specific setup.
+
+If you are enqueing lots of jobs, you could be using [Sidekiq’s bulk enqueing functionality](https://github.com/mperham/sidekiq/wiki/Bulk-Queueing), this is hard to do performantly when using Sidekiq with Active Job.
+
+There are, confusingly, multiple levels of retries in the event of a job failure. Active Job has its own retry mechanism, which once completed then passes down to Sidekiq’s own (seperate and different!) retry system. This is a confusing thing to debug when things go wrong.
```

Comment:
Maybe take out the first 'confusingly' or change the last one to "tricky" or "difficult" or something like that.